### ISSUE TITLE
Recent SVE commit is breaking aarch64 builds on multiple microarchitectures

### ISSUE THREAD

**Issue by Sqvid** [pytorch/pytorch#136440]
### 🐛 Describe the bug

The recent changes introduced by [#119571](https://github.com/pytorch/pytorch/pull/119571) are breaking aarch64 builds on multiple platforms. A similar failure was also reported by [this issue](https://github.com/pytorch/pytorch/issues/136432), however the errors I have observed seem similar, but different and may add further insight into the scope of the issue.

On Neoverse V1:
```
[1195/1202] Linking CXX shared library lib/libtorch_cpu.so                                                                                                                                         
FAILED: lib/libtorch_cpu.so
...
caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/CPUGeneratorImpl.cpp.o: in function `std::__atomic_base<unsigned int>::fetch_add(unsigned int, std::memory_order)':                               
/usr/include/c++/10/bits/atomic_base.h:548:(.text._ZN3c106detail25atomic_refcount_incrementERSt6atomicIjE[_ZN3c106detail25atomic_refcount_incrementERSt6atomicIjE]+0x2c): relocation truncated to f
it: R_AARCH64_CALL26 against symbol `__aarch64_ldadd4_acq_rel' defined in .text section in /usr/lib/gcc/aarch64-linux-gnu/10/libgcc.a(ldadd_4_4.o)                                                 
caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/CPUGeneratorImpl.cpp.o: in function `std::__atomic_base<unsigned int>::fetch_sub(unsigned int, std::memory_order)':                               
/usr/include/c++/10/bits/atomic_base.h:558:(.text._ZN3c106detail25atomic_refcount_decrementERSt6atomicIjE[_ZN3c106detail25atomic_refcount_decrementERSt6atomicIjE]+0x30): relocation truncated to f
it: R_AARCH64_CALL26 against symbol `__aarch64_ldadd4_acq_rel' defined in .text section in /usr/lib/gcc/aarch64-linux-gnu/10/libgcc.a(ldadd_4_4.o)                                                 
/usr/include/c++/10/bits/atomic_base.h:558:(.text._ZN3c106detail26atomic_weakcount_decrementERSt6atomicIjE[_ZN3c106detail26atomic_weakcount_decrementERSt6atomicIjE]+0x30): relocation truncated to
 fit: R_AARCH64_CALL26 against symbol `__aarch64_ldadd4_acq_rel' defined in .text section in /usr/lib/gcc/aarch64-linux-gnu/10/libgcc.a(ldadd_4_4.o)                                               
caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/CPUGeneratorImpl.cpp.o: in function `__gnu_cxx::__exchange_and_add(int volatile*, int)':                                                          
/usr/include/c++/10/ext/atomicity.h:50:(.text._ZNSt16_Sp_counted_baseILN9__gnu_cxx12_Lock_policyE2EE10_M_releaseEv[_ZNSt16_Sp_counted_baseILN9__gnu_cxx12_Lock_policyE2EE10_M_releaseEv]+0x50): rel
ocation truncated to fit: R_AARCH64_CALL26 against symbol `__aarch64_ldadd4_acq_rel' defined in .text section in /usr/lib/gcc/aarch64-linux-gnu/10/libgcc.a(ldadd_4_4.o)                           
/usr/include/c++/10/ext/atomicity.h:50:(.text._ZNSt16_Sp_counted_baseILN9__gnu_cxx12_Lock_policyE2EE10_M_releaseEv[_ZNSt16_Sp_counted_baseILN9__gnu_cxx12_Lock_policyE2EE10_M_releaseEv]+0x104): re
location truncated to fit: R_AARCH64_CALL26 against symbol `__aarch64_ldadd4_acq_rel' defined in .text section in /usr/lib/gcc/aarch64-linux-gnu/10/libgcc.a(ldadd_4_4.o)                          
caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/CachedTensorUtils.cpp.o: in function `std::__atomic_base<unsigned int>::fetch_add(unsigned int, std::memory_order)':                              
/usr/include/c++/10/bits/atomic_base.h:548:(.text._ZN3c106detail26atomic_weakcount_incrementERSt6atomicIjE[_ZN3c106detail26atomic_weakcount_incrementERSt6atomicIjE]+0x28): relocation truncated to
 fit: R_AARCH64_CALL26 against symbol `__aarch64_ldadd4_acq_rel' defined in .text section in /usr/lib/gcc/aarch64-linux-gnu/10/libgcc.a(ldadd_4_4.o)                                               
caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/EmptyTensor.cpp.o: in function `std::__atomic_base<int>::fetch_and(int, std::memory_order)':                                                      
/usr/include/c++/10/bits/atomic_base.h:568:(.text._ZN3c1017SymbolicShapeMeta13refresh_numelEv[_ZN3c1017SymbolicShapeMeta13refresh_numelEv]+0x48): relocation truncated to fit: R_AARCH64_CALL26 aga
inst symbol `__aarch64_ldclr4_acq_rel' defined in .text section in /usr/lib/gcc/aarch64-linux-gnu/10/libgcc.a(ldclr_4_4.o)                                                                         
/usr/include/c++/10/bits/atomic_base.h:568:(.text._ZN3c1017SymbolicShapeMeta18refresh_contiguousEv[_ZN3c1017SymbolicShapeMeta18refresh_contiguousEv]+0x48): relocation truncated to fit: R_AARCH64_
CALL26 against symbol `__aarch64_ldclr4_acq_rel' defined in .text section in /usr/lib/gcc/aarch64-linux-gnu/10/libgcc.a(ldclr_4_4.o)                                                               
caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/FunctionalTensorWrapper.cpp.o: in function `__gnu_cxx::__atomic_add(int volatile*, int)':                                                         
/usr/include/c++/10/ext/atomicity.h:55:(.text._ZNSt16_Sp_counted_baseILN9__gnu_cxx12_Lock_policyE2EE15_M_add_ref_copyEv[_ZNSt16_Sp_counted_baseILN9__gnu_cxx12_Lock_policyE2EE15_M_add_ref_copyEv]+
0x50): relocation truncated to fit: R_AARCH64_CALL26 against symbol `__aarch64_ldadd4_acq_rel' defined in .text section in /usr/lib/gcc/aarch64-linux-gnu/10/libgcc.a(ldadd_4_4.o)                 
caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/FunctionalizeFallbackKernel.cpp.o: in function `std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_add_ref_lock()':                            
/usr/include/c++/10/bits/shared_ptr_base.h:254:(.text._ZNSt16_Sp_counted_baseILN9__gnu_cxx12_Lock_policyE2EE15_M_add_ref_lockEv[_ZNSt16_Sp_counted_baseILN9__gnu_cxx12_Lock_policyE2EE15_M_add_ref_
lockEv]+0x64): relocation truncated to fit: R_AARCH64_CALL26 against symbol `__aarch64_cas4_acq_rel' defined in .text section in /usr/lib/gcc/aarch64-linux-gnu/10/libgcc.a(cas_4_4.o)             
caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/FunctionalizeFallbackKernel.cpp.o: in function `__gnu_cxx::__atomic_add(int volatile*, int)':                                                     
/usr/include/c++/10/ext/atomicity.h:55:(.text._ZNSt16_Sp_counted_baseILN9__gnu_cxx12_Lock_policyE2EE15_M_weak_add_refEv[_ZNSt16_Sp_counted_baseILN9__gnu_cxx12_Lock_policyE2EE15_M_weak_add_refEv]+
0x50): additional relocation overflows omitted from the output                                                                                                                                     
collect2: error: ld returned 1 exit status                                                                                                                                                         
ninja: build stopped: subcommand failed.
```
I should point out on the commit preceding this patch bad69044d87b67a8b64691e4e5a7a68147185aa0, torch builds without an issue.

~Furthermore this patch is breaking Neoverse N1 builds with multiple compilation errors including:~
**EDIT**: this seems to be fixed by https://github.com/pytorch/pytorch/pull/136388
```
/pytorch/aten/src/ATen/cpu/vec/sve/vec_double.h:264:57: error: cannot convert ‘const vls_float64_t’ {aka ‘const svfloat64_t __attribute__((arm_sve_vector_bits(256)))’} to ‘svfloat32_t’
  264 |         return Vectorized<double>(Sleef_nextafterfx_sve(values, b));
      |                                                         ^~~~~~
      |                                                         |
      |                                                         const vls_float64_t {aka const svfloat64_t __attribute__((arm_sve_vector_bits(256)))}

/pytorch/aten/src/ATen/cpu/vec/sve/vec_double.h:9:47: note: in definition of macro ‘USE_SLEEF’
    9 | #define USE_SLEEF(sleef_code, non_sleef_code) sleef_code
      |                                               ^~~~~~~~~~

In file included from /pytorch/aten/src/ATen/cpu/vec/sve/vec_float.h:8,
                 from /pytorch/aten/src/ATen/cpu/vec/sve/vec_common_sve.h:12,
                 from /pytorch/aten/src/ATen/cpu/vec/vec256/vec256.h:11,
                 from /pytorch/aten/src/ATen/cpu/vec/vec.h:6,
                 from /pytorch/aten/src/ATen/native/cpu/Loops.h:37,
                 from /pytorch/aten/src/ATen/native/cpu/spherical_bessel_j0.cpp:8,
                 from /pytorch/build/aten/src/ATen/native/cpu/spherical_bessel_j0.cpp.SVE256.cpp:1:

/pytorch/build/include/sleef.h:1415:60: note:   initializing argument 1 of ‘svfloat32_t Sleef_nextafterfx_sve(svfloat32_t, svfloat32_t)’
 1415 | SLEEF_IMPORT SLEEF_CONST svfloat32_t Sleef_nextafterfx_sve(svfloat32_t, svfloat32_t);
      |                                                            ^~~~~~~~~~~

[5267/6123] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/cpu/scaled_modified_bessel_k1.cpp.SVE256.cpp.o
```

I raised similar concerns [in my PR](https://github.com/pytorch/pytorch/pull/133995).

### Versions
```
Collecting environment information...
PyTorch version: 2.6.0a0+gitd61210f
Is debug build: True
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (aarch64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.12 (main, Sep 11 2024, 15:47:36) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.8.0-1015-aws-aarch64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False

CPU:
Architecture:                         aarch64
CPU op-mode(s):                       32-bit, 64-bit
Byte Order:                           Little Endian
CPU(s):                               64
On-line CPU(s) list:                  0-63
Vendor ID:                            ARM
Model:                                1
Thread(s) per core:                   1
Core(s) per socket:                   64
Socket(s):                            1
Stepping:                             r1p1
BogoMIPS:                             2100.00
Flags:                                fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm jscvt fcma lrcpc dcpop sha3 sm3 sm4 asimddp sha512 sve asimdfhm dit uscat ilrcpc flagm ssbs paca pacg dcpodp svei8mm svebf16 i8mm bf16 dgh rng
L1d cache:                            4 MiB (64 instances)
L1i cache:                            4 MiB (64 instances)
L2 cache:                             64 MiB (64 instances)
L3 cache:                             32 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-63
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; CSV2, BHB
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] flake8==6.1.0
[pip3] flake8-bugbear==23.3.23
[pip3] flake8-comprehensions==3.15.0
[pip3] flake8-executable==2.1.3
[pip3] flake8-logging-format==0.9.0
[pip3] flake8-pyi==23.3.1
[pip3] flake8-simplify==0.19.3
[pip3] mypy==1.10.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.0
[pip3] onnx==1.16.2
[pip3] optree==0.12.1
[pip3] torch==2.6.0a0+gitbad6904
[pip3] torch_geometric==2.4.0
[pip3] torchvision==0.19.0
[conda] Could not collect
```

# Further comments:
I have noticed that after the offending PR I can indeed build PyTorch on `origin/main` if I use `DEBUG=0`. However before the commit both `DEBUG=1` and `DEBUG=0` work. My full build command on Neoverse V1 is (built in a venv with `pip install -r requirements.txt`):

```sh
export CC=${CC:-"gcc-10"}
export CXX=${CXX:-"g++-10"}
export DEBUG=${DEBUG:-"1"}
export USE_MKLDNN=${USE_MKLDNN:-"1"}
export USE_MKLDNN_ACL=${USE_MKLDNN_ACL:-"1"} 
export USE_KINETO=${USE_KINETO:-"0"}
export USE_DISTRIBUTED=${USE_DISTRIBUTED:-"0"}
export USE_NCCL=${USE_NCCL:-"0"}
export BUILD_TEST=${BUILD_TEST:-"0"}
export USE_XNNPACK=${USE_XNNPACK:-"0"}
export USE_FBGEMM=${USE_FBGEMM:-"0"}
export USE_QNNPACK=${USE_QNNPACK:-"0"}
export USE_MIOPEN=${USE_MIOPEN:-"0"}
export USE_NNPACK=${USE_NNPACK:-"0"}
export USE_TENSORPIPE=${USE_TENSORPIPE:-"0"}

python setup.py develop --cmake
```

**Comment by Sqvid** [comment]
@malfet Is there any clarity on how this did not cause failures in the CI build?

**Comment by malfet** [comment]
Assigning to myself to either fwd-fix or disable SVE for now

**Comment by Sqvid** [comment]
> I've fixed vec_double compilation issue by #136388

Yes, sorry I just noticed and was updating my comment as you replied. Please note my observations on debug vs release builds as well.

**Comment by malfet** [comment]
> Yes, sorry I just noticed and was updating my comment as you replied. Please note my observations on debug vs release builds as well.

Very good point, though I wouldn't be surprised if this is the case, and than perhaps trivial fix would be to disable SVE for debug builds (or disable SLEF itnergation)

**Comment by Sqvid** [comment]
> Very good point, though I wouldn't be surprised if this is the case, and than perhaps trivial fix would be to disable SVE for debug builds (or disable SLEF itnergation)

Possibly, though as you may remember I am working on an SVE-related change here as well https://github.com/pytorch/pytorch/pull/133995 which I would like people to be able to test in debug mode if they want to extend it. I feel like the bigger issue is that the breaking commit is forcing SVE into a paradigm it is not designed for. The "S" in SVE stands for **scalable** as the code is supposed to be **vector-length agnostic**. This is the core design principle of SVE. However that entire patch hinges on the concept of specifying different codepaths for different vector lengths, which clashes with the fundamental design-principle of the extension.

**Comment by abhishek-iitmadras** [comment]
> > Yes, sorry I just noticed and was updating my comment as you replied. Please note my observations on debug vs release builds as well.
> 
> Very good point, though I wouldn't be surprised if this is the case, and than perhaps trivial fix would be to disable SVE for debug builds (or disable SLEF itnergation)

Hi @malfet , @Sqvid 

Yes, we need to disable SVE for debug build as temporary fix for this issue.

**Comment by abhishek-iitmadras** [comment]
> > > Yes, sorry I just noticed and was updating my comment as you replied. Please note my observations on debug vs release builds as well.
> > 
> > 
> > Very good point, though I wouldn't be surprised if this is the case, and than perhaps trivial fix would be to disable SVE for debug builds (or disable SLEF itnergation)
> 
> Hi @malfet , @Sqvid
> 
> Yes, we need to disable SVE for debug build as temporary fix for this issue.

This also raise a concern why we don't have any CI that check build with 'export DEBUG=1`

Need to add this type of check in CI especially with ciflow/linux-aarch64

**Comment by maajidkhann** [comment]
> > Very good point, though I wouldn't be surprised if this is the case, and than perhaps trivial fix would be to disable SVE for debug builds (or disable SLEF itnergation)
> 
> Possibly, though as you may remember I am working on an SVE-related change here as well #133995 which I would like people to be able to test in debug mode if they want to extend it. I feel like the bigger issue is that the breaking commit is forcing SVE into a paradigm it is not designed for. The "S" in SVE stands for **scalable** as the code is supposed to be **vector-length agnostic**. This is the core design principle of SVE. However that entire patch hinges on the concept of specifying different codepaths for different vector lengths, which clashes with the fundamental design-principle of the extension.
> 
> Either way, I look forward to hearing about your findings after you investigate the issue. Many thanks.

Hello @Sqvid . The SVE code is VLA and right now we have just added support for SVE256. We have a follow up PR coming to support different VL's. (512 and 128) and it uses the same code.
Right now, we have two flows: SVE and DEFAULT (NEON). The VL is detected at runtime and its corresponding backend is called in dispatch mechanism(DEFAULT or SVE) and in case of SVE, VL detected at runtime is used in the SVE code for calculations making the code VLA.

**Refer to my comment here for more info:**
https://github.com/pytorch/pytorch/pull/134672#issuecomment-2317216527

**Comment by Sqvid** [comment]
@maajidkhann Thanks for your response, I did find your linked comment clarifying. However, (I don't want to derail this discussion too much) as I pointed out on https://github.com/pytorch/pytorch/pull/133995, some of my issues stem from the fact that a lot of code was put in with the idea that it will make sense with future work. Specifically I am thinking of the SVE detection. If we want to discuss it further though it should be on that linked PR.
